### DISCUSSION TITLE
Patch Bundle entry not removing when base setup is uninstalled using v4

### DISCUSSION THREAD

**Discussion by kalyani1217** [wixtoolset/issues#8667] [Questions]
I have created and installed base and HotFix setups.
 

Now, when I uninstall the base setup from ARP, the base setup is removed. However, HF entry still persists in ARP, and files are getting removed. Only HF entry persists in the ARP.

No issue with MSI or MSP. Seeing an issue with Bundle.
 Here are my base and HF bundle files:
Base Bundle.wxs:

```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs" xmlns:bal="http://wixtoolset.org/schemas/v4/wxs/bal">
  <Bundle Name="RTM_Bundle" Manufacturer="TODO Manufacturer" Version="1.0.0.0" UpgradeCode="3d29a1cb-4640-4ea6-b6b8-8597ce22d121">
    <RelatedBundle Id="{1CFF8E41-9AE0-46D9-8E3D-3B555C2465AC}" Action="detect" />
    <BootstrapperApplication>
      <bal:WixStandardBootstrapperApplication LicenseUrl="https://www.example.com/license" Theme="hyperlinkLicense" />
    </BootstrapperApplication>
    <Chain>
      <MsiPackage Id="SmallSetup" Cache="keep" Compressed="yes" Name="SmallSetup" ForcePerMachine="yes"
                   bal:DisplayInternalUICondition="WixBundleAction = 6" EnableFeatureSelection="yes" SourceFile="$(var.SolutionDir)RTM_MSI\bin\x86\Debug\en-US\RTM_MSI.msi"
                   DisplayName="Small Setup " Description="Small Setup " Vital="yes"
                  >
      </MsiPackage>
    </Chain>
  </Bundle>
</Wix>
```

HF Bundle.wxs:

```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs" xmlns:bal="http://wixtoolset.org/schemas/v4/wxs/bal">
  <Bundle Name="HF_Bundle" Manufacturer="TODO Manufacturer" Version="1.0.0.0" UpgradeCode="{56FB724A-2A8D-4BFD-8A92-E58464D5A653}">

    <RelatedBundle Id="{1CFF8E41-9AE0-46D9-8E3D-3B555C2465AC}" Action="patch" />
    <BootstrapperApplication>
      <bal:WixStandardBootstrapperApplication LicenseUrl="https://www.example.com/license" Theme="hyperlinkLicense" />
    </BootstrapperApplication>
    <Chain>
      <MspPackage Id="SmallSetup" Cache="keep" Compressed="yes" Name="SmallSetup" 
                  bal:DisplayInternalUICondition="WixBundleAction = 6" 
                  SourceFile="$(var.SolutionDir)HF_MSP\bin\x86\Debug\HF_MSP.msp"
                  DisplayName="Small Setup" Description="Small Setup" Vital="yes"
                  >
      </MspPackage>
    </Chain>
  </Bundle>
</Wix>
```

The files generated during the uninstallation of the base setup are:
```
Base_Bundle_20240731140954
Base_Bundle_20240731140956.elevated
Base_Bundle_20240731140954_000_SmallSetup
HF_Bundle_20240731140959.elevated
Base_Bundle_20240731140954_001_{3899887C-F6DA-4070-ACC9-15CAA3403B3E}
```
If we observe, there is no log for the HF bundle. 
"Base_Bundle_20240731140954_001_{3899887C-F6DA-4070-ACC9-15CAA3403B3E}" This log is of MSP.

Am I doing anything wrong in Bundle files or why the HF entry is not getting removed?

I would be grateful if you could help me out.

Thanks

**Comment by bevanweiss** [comment]
Where do your RelatedBundle Ids actually point to?
If these are your two bundles, I would expect them to be:
```xml
<Bundle Name="RTM_Bundle" Manufacturer="TODO Manufacturer" Version="1.0.0.0" UpgradeCode="{3d29a1cb-4640-4ea6-b6b8-8597ce22d121}">
   ...
</Bundle>

<Bundle Name="HF_Bundle" Manufacturer="TODO Manufacturer" Version="1.0.0.0" UpgradeCode="{56FB724A-2A8D-4BFD-8A92-E58464D5A653}">
    <RelatedBundle Id="{3d29a1cb-4640-4ea6-b6b8-8597ce22d121}" Action="patch" />
  ...
</Bundle>

```

i.e. your base bundle doesn't have anything related to it, since it's the first bundle (obviously this might change if it's an upgrade or similar from something else)
And then your HF would have the RelatedBundle which points back to the original bundle, and declares itself as a patch to that (i.e. that original bundle needs to be installed first, but they will be uninstalled together if the related bundle is uninstalled).

**Reply by kalyani1217** [reply]
Thanks @bevanweiss for your response.
I have tried the following steps:

Removed the RelatedBundle element in the base setup.
Set the UpgradeCode of the base setup to the RelatedBundle ID of the hotfix, and Action to Patch.
However, **I'm still observing the same issue**. The bundle entry of the hotfix is not removed when I uninstall the base setup.

Here are the bundle files.
Base Bundle.wxs:
```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs" xmlns:bal="http://wixtoolset.org/schemas/v4/wxs/bal">
  <Bundle Name="RTM_Bundle" Manufacturer="TODO Manufacturer" Version="1.0.0.0" UpgradeCode="{3d29a1cb-4640-4ea6-b6b8-8597ce22d121}">
    <BootstrapperApplication>
      <bal:WixStandardBootstrapperApplication LicenseUrl="https://www.example.com/license" Theme="hyperlinkLicense" />
    </BootstrapperApplication>
    <Chain>
      <MsiPackage Id="SmallSetup" Cache="keep" Compressed="yes" Name="SmallSetup" ForcePerMachine="yes"
                   bal:DisplayInternalUICondition="WixBundleAction = 6" EnableFeatureSelection="yes" SourceFile="$(var.SolutionDir)RTM_MSI\bin\x86\Debug\en-US\RTM_MSI.msi"
                   DisplayName="Small Setup Package DisplayName" Description="Small Setup Package Description" Vital="yes"
                  >
      </MsiPackage>
    </Chain>
  </Bundle>
</Wix>
```

HF Bundle.wxs:
```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs" xmlns:bal="http://wixtoolset.org/schemas/v4/wxs/bal">
  <Bundle Name="HF_Bundle" Manufacturer="TODO Manufacturer" Version="1.0.0.101" UpgradeCode="{56FB724A-2A8D-4BFD-8A92-E58464D5A653}">
    <RelatedBundle Id="{3d29a1cb-4640-4ea6-b6b8-8597ce22d121}" Action="patch" />
    <BootstrapperApplication>
      <bal:WixStandardBootstrapperApplication LicenseUrl="https://www.example.com/license" Theme="hyperlinkLicense" />
    </BootstrapperApplication>
    <Chain>
      <MspPackage Id="SmallSetup" Cache="keep" Compressed="yes" Name="SmallSetup" 
                  bal:DisplayInternalUICondition="WixBundleAction = 6" 
                  SourceFile="$(var.SolutionDir)HF_MSP\bin\x86\Debug\HF_MSP.msp"
                  DisplayName="Small Setup Package DisplayName" Description="Small Setup Package Description" Vital="yes"
                  >
      </MspPackage>
    </Chain>
  </Bundle>
</Wix>
```

**Comment by kalyani1217** [comment]
I even tested with both BundleJ and BundleJ_Patch, located at [wix/src/test/burn/TestData/DependencyTests](https://github.com/wixtoolset/wix/tree/ce73352b1fa1d4f9cded10a0ee410f2e786bd326/src/test/burn/TestData/DependencyTests)
Observed the same issue with these setups as well: after uninstalling the BundleJ setup, patch entry still persists in the registry.

Attached log files were generated during uninstallation of base setup i.e., BundleJ

[~DependencyTests_BundleJ_20240801123804.log](https://github.com/user-attachments/files/16451682/DependencyTests_BundleJ_20240801123804.log)
[~DependencyTests_BundleJ_20240801123804_000_PackageA.log](https://github.com/user-attachments/files/16451683/DependencyTests_BundleJ_20240801123804_000_PackageA.log)
[~DependencyTests_BundleJ_20240801123809.elevated.log](https://github.com/user-attachments/files/16451684/DependencyTests_BundleJ_20240801123809.elevated.log)
[~DependencyTests_BundleJ_Patch_20240801123810.cleanroom.log](https://github.com/user-attachments/files/16451685/DependencyTests_BundleJ_Patch_20240801123810.cleanroom.log)
[~DependencyTests_BundleJ_Patch_20240801123811.log](https://github.com/user-attachments/files/16451686/DependencyTests_BundleJ_Patch_20240801123811.log)
[~DependencyTests_BundleJ_Patch_20240801123811_000_PackageE.log](https://github.com/user-attachments/files/16451687/DependencyTests_BundleJ_Patch_20240801123811_000_PackageE.log)
[~DependencyTests_BundleJ_Patch_20240801123817.elevated.log](https://github.com/user-attachments/files/16451688/DependencyTests_BundleJ_Patch_20240801123817.elevated.log)


Thanks

**Reply by kalyani1217** [reply]
@bevanweiss  and @robmen ,Could you please look into this issue.

**Reply by robmen** [reply]
Certainly. I provide [dedicated support for the WiX Toolset for companies of all shapes and sizes](https://www.firegiant.com/services/).

**Reply by kalyani1217** [reply]
@robmen, It seems to be a bug with WiX. The HF log indicates that the package install registration state is present. 

As the state of the MSP package install registration state is present, which causes the resume type to be set to ARP. As a result, the bundle was not removed because the resume type is ARP.

Why does the MSP package install registration state is present even though it is being uninstalled when the base setup is uninstalled?

**[12714:17328][2024-08-01T12:38:19]i374:  package: PatchA, install registration state: Present, cache registration state: Present**
[12714:17328][2024-08-01T12:38:19]i374:   package: PackageE, install registration state: Absent, cache registration state: Absent
[12714:17328][2024-08-01T12:38:19]i000: TESTBA (patch): OnUnregisterBegin, default: Full, requested: Full
[12714:17328][2024-08-01T12:38:19]i372: Session end, registration key: SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{2777CFCE-44DF-4F6B-85A0-E9D6B3143992}, **resume: ARP**, restart: None, disable resume: Yes, default registration: Full, ba requested registration: Full
[12A24:CCDC][2024-08-01T12:38:19]i371: Updating session, registration key: SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{2777CFCE-44DF-4F6B-85A0-E9D6B3143992}, resume: ARP, restart initiated: No, disable resume: Yes

**Reply by bevanweiss** [reply]
`<sarcasm>`
Ahhh... a bug.  You will be pleased to know that the WiX Toolset has a 100% money back guarantee for anyone that uncovers a bug...  On presentation of the invoice for your purchase of the official WiX Toolset, you will be refunded the full invoice value.
`</sarcasm>`

Since it's all open source, you do have the capability of fixing the bug yourself.
If it's urgent to you, or you have a financial interest in it, then your best option is to contact FireGiant for support.

If you do have an absolute minimal reproducible example of deviation from documented behaviour, then you could raise an issue with the repro repo.  It's still not a guarantee of a fix, and even if it is potentially fixed, there's no guarantee of a timeline for such a fix.

**Reply by robmen** [reply]
Note: @bevanweiss is correct. Without sarcasm, [FireGiant offers guaranteed support for the WiX Toolset](https://www.firegiant.com/services/support-programs/#guaranteed-wix-bug-fixes-enterprise-only).

Note: we do try really hard to not have bugs in the WiX Toolset but they can happen, especially in complex interactions like bundle patching and upgrades.
